rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -247,11 +260,126 @@ public void close() {
         try {
             flush();
             base.close();
+            if (cacheSnapshotPath != null) {
+                saveCommittedCache(unsafeCommittedCache, cacheSnapshotPath);
+            }
             uncommittedCache.clear();
             committedCache.clear();
         }
         finally {
             this.lock.writeLock().unlock();
         }
     }
+
+    private static Map<ByteArrayWrapper, byte[]> makeCommittedCache(int cacheSize, @Nullable Path cacheSnapshotPath) {
+        Map<ByteArrayWrapper, byte[]> cache = new MaxSizeHashMap<>(cacheSize, true);
+        if (cacheSnapshotPath == null) {
+            return cache;
+        }
+
+        File cacheSnapshotFile = cacheSnapshotPath.toFile();
+        if (!cacheSnapshotFile.exists()) {
+            return cache;
+        }
+
+        try (DataInputStream inStream = new DataInputStream(new BufferedInputStream(new FileInputStream(cacheSnapshotFile)))) {
+            int count = inStream.readInt();
+            if (count < 1) {
+                throw new IOException("Invalid data: number of entries");
+            }
+            int readSize;
+            for (int i = 0; i < count; i++) {
+                int keySize = inStream.readInt();
+                if (keySize < 1) {
+                    throw new IOException("Invalid data: key size");
+                }
+                byte[] key = new byte[keySize];
+                readSize = inStream.read(key);
+                if (readSize != keySize) {
+                    throw new IOException("Invalid data: actual key length");
+                }
+
+                int valueSize = inStream.readInt();
+                if (valueSize < -1) {
+                    throw new IOException("Invalid data: value size");
+                }
+                byte[] value = null;
+                if (valueSize > 0) {
+                    value = new byte[valueSize];
+                    readSize = inStream.read(value);
+                    if (readSize != valueSize) {
+                        throw new IOException("Invalid data: actual value length");
+                    }
+                } else if (valueSize == 0) {
+                    value = new byte[0];
+                }
+
+                cache.put(ByteUtil.wrap(key), value);
+            }
+        } catch (IOException e) {
+            cache.clear();
+            boolean delResult = cacheSnapshotFile.delete();
+            logger.error("Cannot read from cache snapshot file. File deletion result: '{}'", delResult, e);
```

Comment:
should it be deleted on error? Maybe it's better to rename it. For example, add the suffix ".err" to its name. Then it's possible to send the file to more advanced users for debugging, or debug locally by renaming the file back.